R/naive_direction_forecaster.R

```diff
@@ -0,0 +1,163 @@
+#
+
+library(checkmate)
+library(dplyr)
+library(epidatr)
+library(epiprocess)
+library(readr)
+library(rlang)
+library(tibble)
+library(tidyr)
+source(here::here("R", "approx-cdf.R"))
+source(here::here("R", "utils.R"))
+
+
+get_direction_predictions <- function(
+    forecast_due_date,
+    reference_date,
+    quantile_predictions) {
+  augmented_location_data <- get_flusight_location_data()
+  state_pop <- get_state_data()
+
+  # We need to combine the quantile forecasts with recent observations in order to
+  # do direction calculations.
+  # TODO: Using a workaround for https://github.com/cmu-delphi/epidatr/issues/194
+  short_snapshot <- bind_rows(
+    epidatr::pub_covidcast(
+      "hhs",
+      "confirmed_admissions_influenza_1d",
+      "state",
+      "day",
+      "*",
+      epirange(forecast_due_date - 20L, forecast_due_date),
+      as_of = strftime(forecast_due_date, "%Y-%m-%d")
+    ),
+    epidatr::pub_covidcast(
+      "hhs",
+      "confirmed_admissions_influenza_1d",
+      "nation",
+      "day",
+      "*",
+      epirange(forecast_due_date - 20L, forecast_due_date),
+      as_of = strftime(forecast_due_date, "%Y-%m-%d")
+    ) %>% as_tibble()
+  )
+
+  # By the cadence of HHS data, we expect to have data up to Friday 12 days back
+  # from the forecast due date. We get the latest 7 day sum of the data available.
+  reference_7d_counts <- short_snapshot %>%
+    filter(time_value <= forecast_due_date - 12L) %>%
+    group_by(geo_value) %>%
+    complete(time_value = full_seq(time_value, 1L)) %>%
+    slice_max(time_value, n = 7L) %>%
+    summarize(
+      reference_7dcount = sum(value),
+      time_value = max(time_value),
+      .groups = "drop"
+    )
+
+  # Rule of three for probability of "novel behavior"; mixing weight for uniform
+  uniform_forecaster_weight <- 3 / (
+    # roughly, how many years of hhs influenza are available
+    as.numeric(forecast_due_date - as.Date("2020-10-16")) / (365 + 1 / 4 - 1 / 100 + 1 / 400) *
+      # say there's one wave/season per year
+      1 *
+      # say states from separate HHS Regions are "separate" enough to count as different data points, while those within the same region aren't separate enough; 10 HHS Regions = 10 buckets
+      10
+  )
+
+  direction_predictions <- quantile_predictions %>%
+    group_by(
+      forecaster,
+      data_source,
+      signal,
+      geo_value,
+      incidence_period,
+      reference_date,
+      target,
+      horizon,
+      target_end_date,
+      location
+    ) %>%
+    summarize(
+      forecast_acdf = list(approx_cdf_from_quantiles(value, output_type_id)),
+      .groups = "keep"
+    ) %>%
+    left_join(augmented_location_data %>% select(-geo_value), by = "location") %>%
+    left_join(reference_7d_counts, by = "geo_value") %>%
+    reframe(
+      output_type = "pmf",
+      output_type_id = c("large_decrease", "decrease", "stable", "increase", "large_increase"),
+      value = {
+        stopifnot(length(forecast_acdf) == 1L)
```

Comment:
thought: Wish there were some base/tidyverse `unwrap` function in R; would be nice here.  Haven't found one yet.  You might be able to use `rowwise()` and omit the `[[1L]]`'s on list columns?  Don't quite remember.